--- mvpdist/kernel/drivers/Makefile	2003-11-13 17:14:34.000000000 -0500
+++ mvpkdev/kernel/drivers/Makefile	2005-10-16 16:20:31.000000000 -0400
@@ -9,12 +9,13 @@
 mod-subdirs :=	dio mtd sbus video macintosh usb input telephony sgi ide \
 		message/i2o message/fusion scsi md ieee1394 pnp isdn atm \
 		fc4 net/hamradio i2c l3 acpi bluetooth serial maple vme \
-		sensors
+		sensors mvpstb
 
 subdir-y :=	parport char block net sound misc media cdrom hotplug
 subdir-m :=	$(subdir-y)
 
 
+subdir-$(CONFIG_MVPSTB_MOD)		+= mvpstb
 subdir-$(CONFIG_DIO)		+= dio
 subdir-$(CONFIG_PCI)		+= pci
 subdir-$(CONFIG_PCMCIA)		+= pcmcia
--- mvpdist/kernel/arch/ppc/config.in	2003-11-13 17:14:34.000000000 -0500
+++ mvpkdev/kernel/arch/ppc/config.in	2005-10-16 16:20:07.000000000 -0400
@@ -675,4 +675,9 @@
 fi
 endmenu
 
+mainmenu_option next_comment
+comment 'MVP STB module'
+tristate 'MediaMVP Custom STB support' CONFIG_MVPSTB_MOD
+endmenu
+
 source lib/Config.in
