--- mvpdist/kernel/kernel/printk.c.orig	2005-01-16 16:45:10.017850229 -0500
+++ mvpdist/kernel/kernel/printk.c	2005-01-16 18:00:31.776182062 -0500
@@ -47,7 +47,7 @@
 
 /* We show everything that is MORE important than this.. */
 #define MINIMUM_CONSOLE_LOGLEVEL 1 /* Minimum loglevel we let people use */
-#define DEFAULT_CONSOLE_LOGLEVEL 7 /* anything MORE serious than KERN_DEBUG */
+#define DEFAULT_CONSOLE_LOGLEVEL 1 /* Don't send anything to console */
 
 DECLARE_WAIT_QUEUE_HEAD(log_wait);
 
