diff -ur buildroot.dist/toolchain/kernel-headers/kernel-headers.mk buildroot/toolchain/kernel-headers/kernel-headers.mk
--- buildroot.dist/toolchain/kernel-headers/kernel-headers.mk	2006-07-10 08:15:05.000000000 +0100
+++ buildroot/toolchain/kernel-headers/kernel-headers.mk	2009-07-14 22:17:38.000000000 +0100
@@ -19,7 +19,7 @@
 VERSION:=2
 PATCHLEVEL:=4
 SUBLEVEL:=25
-LINUX_HEADERS_SITE:=http://www.uclibc.org/downloads/toolchain
+LINUX_HEADERS_SITE:=http://www.uclibc.org/downloads/old-releases/toolchain
 LINUX_HEADERS_SOURCE:=linux-libc-headers-2.4.25.tar.bz2
 LINUX_HEADERS_UNPACK_DIR:=$(TOOL_BUILD_DIR)/linux-libc-headers-2.4.25
 endif
@@ -28,7 +28,7 @@
 VERSION:=2
 PATCHLEVEL:=4
 SUBLEVEL:=27
-LINUX_HEADERS_SITE:=http://www.uclibc.org/downloads/toolchain
+LINUX_HEADERS_SITE:=http://www.uclibc.org/downloads/old-releases/toolchain
 LINUX_HEADERS_SOURCE:=linux-libc-headers-2.4.27.tar.bz2
 LINUX_HEADERS_UNPACK_DIR:=$(TOOL_BUILD_DIR)/linux-libc-headers-2.4.27
 endif
@@ -37,7 +37,7 @@
 VERSION:=2
 PATCHLEVEL:=4
 SUBLEVEL:=29
-LINUX_HEADERS_SITE:=http://www.uclibc.org/downloads/toolchain
+LINUX_HEADERS_SITE:=http://www.uclibc.org/downloads/old-releases/toolchain
 LINUX_HEADERS_SOURCE:=linux-libc-headers-2.4.29.tar.bz2
 LINUX_HEADERS_UNPACK_DIR:=$(TOOL_BUILD_DIR)/linux-libc-headers-2.4.29
 endif
@@ -46,7 +46,7 @@
 VERSION:=2
 PATCHLEVEL:=4
 SUBLEVEL:=31
-LINUX_HEADERS_SITE:=http://www.uclibc.org/downloads/toolchain
+LINUX_HEADERS_SITE:=http://www.uclibc.org/downloads/old-releases/toolchain
 LINUX_HEADERS_SOURCE:=linux-libc-headers-2.4.31.tar.bz2
 LINUX_HEADERS_UNPACK_DIR:=$(TOOL_BUILD_DIR)/linux-libc-headers-2.4.31
 endif
