--- uClibc-0.9.23/libc/sysdeps/linux/powerpc/setjmp.S	2003-07-03 05:40:18.000000000 -0500
+++ uClibc-0.9.23/libc/sysdeps/linux/powerpc/setjmp.S	2004-08-28 18:22:03.000000000 -0500
@@ -23,7 +23,7 @@
 #define _SETJMP_H
 #include <bits/setjmp.h>
 
-#ifdef __UCLIBC_HAS_FLOATS__ && ! defined __UCLIBC_HAS_SOFT_FLOAT__
+#if defined __UCLIBC_HAS_FLOATS__ && ! defined __UCLIBC_HAS_SOFT_FLOAT__
 #define FP(x...) x
 #else
 #define FP(x...)
