diff -Narup linux-2.4.31-orig/Makefile linux-2.4.31/Makefile
--- linux-2.4.31-orig/Makefile	2005-08-09 11:30:57.000000000 -0700
+++ linux-2.4.31/Makefile	2005-08-09 11:33:15.000000000 -0700
@@ -5,7 +5,7 @@ EXTRAVERSION =
 
 KERNELRELEASE=$(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)
 
-ARCH := $(shell uname -m | sed -e s/i.86/i386/ -e s/sun4u/sparc64/ -e s/arm.*/arm/ -e s/sa110/arm/)
+#ARCH := $(shell uname -m | sed -e s/i.86/i386/ -e s/sun4u/sparc64/ -e s/arm.*/arm/ -e s/sa110/arm/)
 KERNELPATH=kernel-$(shell echo $(KERNELRELEASE) | sed -e "s/-//g")
 
 CONFIG_SHELL := $(shell if [ -x "$$BASH" ]; then echo $$BASH; \
@@ -19,7 +19,10 @@ FINDHPATH	= $(HPATH)/asm $(HPATH)/linux 
 HOSTCC  	= gcc
 HOSTCFLAGS	= -Wall -Wstrict-prototypes -O2 -fomit-frame-pointer
 
-CROSS_COMPILE 	=
+#CROSS_COMPILE 	=
+
+ARCH=ppc
+CROSS_COMPILE	= $(CROSS)
 
 #
 # Include the make variables (CC, etc...)
