--- linux-2.4.31/include/asm-ppc/processor.h	2005-04-03 20:42:20.000000000 -0500
+++ linux-2.4.31-new/include/asm-ppc/processor.h	2006-11-26 17:15:23.000000000 -0600
@@ -880,12 +882,14 @@
 /*
  * NOTE! The task struct and the stack go together
  */
-#define THREAD_SIZE (2*PAGE_SIZE)
+#define THREAD_SIZE (3*PAGE_SIZE)
 #define alloc_task_struct() \
-	((struct task_struct *) __get_free_pages(GFP_KERNEL,1))
-#define free_task_struct(p)	free_pages((unsigned long)(p),1)
+	((struct task_struct *) __get_free_pages(GFP_KERNEL,get_order(THREAD_SIZE)))
+#define free_task_struct(p)	free_pages((unsigned long)(p),get_order(THREAD_SIZE))
 #define get_task_struct(tsk)      atomic_inc(&virt_to_page(tsk)->count)
 
+#define INIT_TASK_SIZE THREAD_SIZE
+
 /* in process.c - for early bootup debug -- Cort */
 int ll_printk(const char *, ...);
 void ll_puts(const char *);
