diff -Narup linux-2.4.31-orig/arch/ppc/boot/utils/Makefile linux-2.4.31/arch/ppc/boot/utils/Makefile
--- linux-2.4.31-orig/arch/ppc/boot/utils/Makefile	2003-06-13 07:51:31.000000000 -0700
+++ linux-2.4.31/arch/ppc/boot/utils/Makefile	2005-08-03 06:08:06.000000000 -0700
@@ -5,7 +5,7 @@
 # License.  See the file "COPYING" in the main directory of this archive
 # for more details.
 
-HOSTCFLAGS += -I$(TOPDIR)/arch/$(ARCH)/boot/include
+HOSTCFLAGS += -I$(TOPDIR)/arch/$(ARCH)/boot/include -m32
 
 all: dummy
 
