--- microwindows-0.91/src/config	2003-11-16 16:34:12.000000000 -0600
+++ mw/src/config	2006-03-13 21:55:13.000000000 -0600
@@ -38,8 +38,7 @@
 #
 ####################################################################
 ARCH                     = LINUX-NATIVE
-BIGENDIAN                = N
-NATIVETOOLSPREFIX        = 
+BIGENDIAN                = Y
 ARMTOOLSPREFIX           = arm-linux-
 MIPSTOOLSPREFIX          = mipsel-linux-
 POWERPCTOOLSPREFIX       = powerpc-linux-
@@ -52,9 +51,9 @@
 #
 ####################################################################
 OPTIMIZE                 = Y
-DEBUG                    = N
+DEBUG                    = Y
 VERBOSE                  = N
-THREADSAFE               = Y
+THREADSAFE               = N
 GPROF                    = N
 
 ####################################################################
@@ -64,7 +63,7 @@
 ####################################################################
 MICROWIN                 = Y
 NANOX                    = Y
-SHAREDLIBS               = Y
+SHAREDLIBS               = N
 OBJFRAMEWORK             = N
 
 
@@ -108,7 +107,7 @@
 # or for speed or debugging.  This affects the nano-X server only.
 #
 ####################################################################
-LINK_APP_INTO_SERVER     = N
+LINK_APP_INTO_SERVER     = Y
 
 ####################################################################
 # Shared memory support for Nano-X client/server protocol speedup
@@ -136,13 +135,14 @@
 ####################################################################
 HAVE_JPEG_SUPPORT        = Y
 INCJPEG                  = .
-LIBJPEG                  = /usr/lib/libjpeg.so
+LIBJPEG                  = /usr/lib/libjpeg.a
 
 ####################################################################
 # PNG support via libpng and libz
 ####################################################################
-HAVE_PNG_SUPPORT         = N
-INCPNG                   = .
+HAVE_PNG_SUPPORT         = Y
+INCPNG                   = /usr/include
+INCLIBZ                   = /usr/include
 LIBPNG                   = /usr/lib/libpng.a
 LIBZ                     = /usr/lib/libz.a
 
@@ -150,14 +150,14 @@
 # TIFF support through libtiff
 ####################################################################
 HAVE_TIFF_SUPPORT        = N
-INCTIFF                  = .
+INCTIFF                  = /usr/include
 LIBTIFF                  = /usr/lib/libtiff.a
 
 ####################################################################
 # native .fnt loadable font support
 ####################################################################
 HAVE_FNT_SUPPORT         = Y
-HAVE_FNTGZ_SUPPORT       = Y
+HAVE_FNTGZ_SUPPORT       = N
 FNT_FONT_DIR             = "fonts/bdf"
 
 ####################################################################
@@ -165,13 +165,13 @@
 ####################################################################
 HAVE_T1LIB_SUPPORT       = N
 INCT1LIB                 = /usr/include
-LIBT1LIB                 = /usr/local/lib/libt1.a
+LIBT1LIB                 = /usr/lib/libt1.a
 
 ####################################################################
 # TrueType font support thru FreeType 1.x
 ####################################################################
-HAVE_FREETYPE_SUPPORT    = Y
-INCFTLIB                 = /usr/include/freetype1
+HAVE_FREETYPE_SUPPORT    = N
+INCFTLIB                 = /usr/include
 LIBFTLIB                 = /usr/lib/libttf.so
 FREETYPE_FONT_DIR        = "fonts/truetype"
 
@@ -188,8 +188,8 @@
 # Selecting HAVE_PCFGZ_SUPPORT will allow you to directly read
 # .pcf.gz files, but it will add libz to the size of the server
 ####################################################################
-HAVE_PCF_SUPPORT         = Y
-HAVE_PCFGZ_SUPPORT       = Y
+HAVE_PCF_SUPPORT         = N
+HAVE_PCFGZ_SUPPORT       = N
 PCF_FONT_DIR             = "fonts/pcf"
 
 ####################################################################
@@ -257,11 +257,11 @@
 ifneq ($(ARCH), ELKS)
 
 # X Window screen, mouse and kbd drivers
-X11                      = N
+X11                      = Y
 
 ifeq ($(X11), Y)
 # startup screen width, height, (depth for palette mode only)
-SCREEN_WIDTH             = 640
+SCREEN_WIDTH             = 720
 SCREEN_HEIGHT            = 480
 SCREEN_DEPTH             = 4
 
@@ -326,7 +326,6 @@
 PSIONMOUSE               = N
 YOPYMOUSE                = N
 HARRIERMOUSE             = N
-LIRCMOUSE                = N
 
 # keyboard or null kbd driver
 TTYKBD                   = N
@@ -338,11 +337,6 @@
 
 endif
 
-# Secondary keyboard drivers.
-# You may have a normal keyboard driver in addition to these
-# drivers, e.g. for both normal keyboard and IR input.
-LIRCKBD2                 = N
-
 ####################################################################
 # Screen driver specific configuration
 # SA1100_LCD_LTLEND 4bpp driver with arm SA1100 LCD controller
