diff --git a/src/Makefile b/src/Makefile
index 6e84b0a..b9ea8e9 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -37,7 +37,8 @@ all: default
 ifeq ($(ARCH), RTEMS)
 	$(MAKE) -C $(MW_DIR_SRC)/rtems
 endif
-	-$(MAKE) -C $(MW_DIR_SRC)/demos
+#Disable demos because they don't currently build correctly for the MVP
+#	-$(MAKE) -C $(MW_DIR_SRC)/demos
 ifeq ($(ARCH), ECOS)
 	$(MAKE) -C $(MW_DIR_SRC)/ecos
 endif
